### DISCUSSION TITLE
How to push policy bundle to OPA

### DISCUSSION THREAD

**Discussion by dushyant7917** [open-policy-agent/community#320] [OPA and Rego] [answered]
In the OPA [documentation](https://www.openpolicyagent.org/docs/latest/management-bundles/) it's mentioned that policy bundle(s) can be fetched from a service periodically or it can be pushed into the OPA using some REST api. But how exactly can we push a specific policy into OPA side car through REST api? Note I want to push/replace/override a complete policy bundle not just a sub-part of it.

**Comment by charlieegan3** [comment, marked as answer]
Hi @dushyant7917 👋

While I'm not sure this is what is being referenced in that page of the docs, it is possible to use HTTP long polling to get faster policy updates to an OPA instance.

This functionality is documented here: https://www.openpolicyagent.org/docs/latest/management-bundles/#http-long-polling. Note however that this requires the bundle service to support the feature (via sending the `application/vnd.openpolicyagent.bundles` Content Type header).

> Note I want to push/replace/override a complete policy bundle

This would support this too.

Does that sound like something that could work for you?

**Reply by dushyant7917** [reply]
There is something mentioned about push model [here](https://www.openpolicyagent.org/docs/latest/management-bundles/)(in the starting section only). `Bundles provide an alternative to pushing policies into OPA via the REST APIs.` This has confused me. So you are saying that OPA doesn't support push model at all? There is no way we can push policies into OPA. We can only pull policies (long/short polling) from a remote server?

**Reply by anderseknert** [reply]
The REST API is documented [here](https://www.openpolicyagent.org/docs/latest/rest-api/), including docs for how to push a policy into OPA.

**Reply by dushyant7917** [reply]
> The REST API is documented [here](https://www.openpolicyagent.org/docs/latest/rest-api/), including docs for how to push a policy into OPA.

Policy API is different. It's meant only for single *.rego file not policy bundle (a dir containing multiple rego files and json data).

**Reply by anderseknert** [reply]
Ah, gotcha! Yeah, you can't push bundles into OPA — those are only consumed by OPA (pull based) via the bundle API.
